### PR TITLE
Startup/shutdown curve

### LINKED ISSUE #221: Testing for BaseMVA scaling
- [ ] scale/unscale -> attributes should be the same
- [ ] scale -> spot check various values (some that should be scaled, others that should not be) -> unscale -> spot check various values
- [ ] regression: scaled / unscaled versions of unit commitment should yield the same answer (easy to change the baseMVA for basic UC models).

Thoughts on using `pytest` for nested dictionaries with almost equal values: https://stackoverflow.com/questions/56046524/check-if-python-dictionaries-are-equal-allowing-small-difference-for-floats

### PR DESCRIPTION
**Pull request by bknueven**
Adds the capability to specify start-up and shut-down power curves for thermal generators within the unit commitment models, which specify power output just before turning on and just after turning off. This is especially important for models with short (e.g., 5-minute) time-horizons.

Curves are specified through the `startup_curve` and `shutdown_curve` attribute on a thermal generator. These represent, from highest to lowest, the power output of the generator in the time periods just before / just after starting-up or shutting-down. Also adds a `future_status` parameter for start-ups after the end of the time horizon. An example is provided in the test `tiny_uc_11.json`.

Also makes some significant changes to the scaling/unscaling functions. Resolves #221 by adding some basic tests for baseMVA scaling/unscaling.

### CODE REVIEW

File: egret/model_library/unit_commitment/power_vars.py [hunk 1]
```diff
@@ -20,6 +20,53 @@ def reactive_power_bounds_rule(m,g,t):
...
+            return linear_expr(linear_vars=linear_vars, linear_coefs=linear_coefs)
+
+    else:
+        def power_generated_grid_expr_rule(m, g, t):
+            startup_curve = m.StartupCurve[g]
+            shutdown_curve = m.ShutdownCurve[g]
+            time_periods_before_startup = value(m.TimePeriodsBeforeStartup[g])
+            time_periods_since_shutdown = value(m.TimePeriodsSinceShutdown[g])
+
+            future_past_production = 0.
```
Comment (jeanpaulwatson) [inline egret/model_library/unit_commitment/power_vars.py:48]: This is a very confusing variable name - I suggest renaming.

Comment (jeanpaulwatson) [reply egret/model_library/unit_commitment/power_vars.py:48]: And no, I don't have a good recommendation.

File: egret/model_library/unit_commitment/params.py [hunk 2]
```diff
@@ -524,12 +524,61 @@ def t0_unit_on_rule(m, g):
     
     _add_initial_time_periods_on_off_line(model)
     _verify_must_run_t0_state_consistency(model)
+
+    # For future shutdowns/startups beyond the time-horizon
+    model.FutureStatus = Param(model.ThermalGenerators,
```
[the remark this answers is not in the document]
Comment (jeanpaulwatson) [reply egret/model_library/unit_commitment/params.py:529]: Basically, that it's the future analog of T0State.

File: egret/model_library/unit_commitment/params.py [hunk 3]
```diff
@@ -524,12 +524,61 @@ def t0_unit_on_rule(m, g):
...
+            return int(math.ceil( value(m.FutureStatus[g]) / value(m.TimePeriodLengthHours) ))
+    model.TimePeriodsBeforeStartup = Param(model.ThermalGenerators, within=PositiveIntegers, mutable=True,
+                                            initialize=time_periods_before_startup_rule)
+
+    ###############################################
+    # startup/shutdown curves for each generator. #
+    ###############################################
+
+    def startup_curve_init_rule(m,g):
+        startup_curve = thermal_gens[g].get('startup_curve')
```
Comment (jeanpaulwatson) [inline egret/model_library/unit_commitment/params.py:558]: The startup_curve quantities have to be in the same time resolution as the scaled min_down_time, as opposed to the MinimumDownTime. Probably worth documenting that requirement somewhere. Although I'm not sure where. Ditto for shutdown_curve analogs.

File: egret/model_library/unit_commitment/power_vars.py [hunk 4]
```diff
@@ -20,6 +20,53 @@ def reactive_power_bounds_rule(m,g,t):
         return (m.MinimumReactivePowerOutput[g,t], m.MaximumReactivePowerOutput[g,t])
     model.ReactivePowerGenerated = Var(model.ThermalGenerators, model.TimePeriods, within=Reals, bounds=reactive_power_bounds_rule)
 
+def _add_power_generated_grid(model):
```
Comment (jeanpaulwatson) [inline egret/model_library/unit_commitment/power_vars.py:23]: I don't like the "_grid" suffix name, but again don't have a better term...

File: egret/model_library/unit_commitment/power_vars.py [hunk 5]
```diff
@@ -20,6 +20,53 @@ def reactive_power_bounds_rule(m,g,t):
...
+            for startup_idx in range(1, min( len(startup_curve)+1, m.NumTimePeriods+1-t )):
+                linear_vars.append(m.UnitStart[g,t+startup_idx])
+                linear_coefs.append(startup_curve[startup_idx])
+            for shutdown_idx in range(1, min( len(shutdown_curve)+1, t+1 )):
+                linear_vars.append(m.UnitStop[g,t-shutdown_idx+1])
+                linear_coefs.append(shutdown_curve[shutdown_idx])
+            linear_expr = get_linear_expr(m.UnitOn, m.UnitStart, m.UnitStop)
+            return linear_expr(linear_vars=linear_vars, linear_coefs=linear_coefs, constant=future_past_production)
+
+    model.PowerGeneratedStartupShutdown = Expression(model.ThermalGenerators, model.TimePeriods,
```
Comment (jeanpaulwatson) [inline egret/model_library/unit_commitment/power_vars.py:67]: The scoping for the power_generated_grid_expr_rule is kind of too cute - it's out of scope in anything other than Python by the time one gets here. Slightly less cute would be to create a variable at this scope and assign the function based on the conditional.

**Review by jeanpaulwatson — APPROVED** [review]
Nothing really required associated with my comments - all just suggestions for clarity improvement.

File: egret/model_library/unit_commitment/power_vars.py (hunk 5 above)
Comment (bknueven) [reply egret/model_library/unit_commitment/power_vars.py:68]: I considered several alternatives -- I'm not sure I like any better than this.

File: egret/model_library/transmission/tx_utils.py [hunk 6]
```diff
@@ -436,7 +438,12 @@ def _scale_by_baseMVA(normal_op, inverse_op, element, attr_name, attr, baseMVA,
                 _scale_by_baseMVA(normal_op, inverse_op, attr, k, v, baseMVA, attributes)
     elif attr_name in attributes:
         op = _get_op(normal_op, inverse_op, attr_name)
-        element[attr_name] = op( attr , baseMVA )
+        if isinstance(attr, list):
```
Comment (jeanpaulwatson) [inline egret/model_library/transmission/tx_utils.py:441]: I am convinced that a blackout will happen in the near future due to failure to apply scaling in some deep, dark corner of an MMS/EMS...

**Review by jeanpaulwatson — APPROVED** [review]

**Comment by bknueven** [comment]
There's some weirdness in scaling/unscaling I'd like to resolve before this is merged

**Pull request merged**